### PR TITLE
bulk update

### PR DESCRIPTION
**Pull request by ereteog**
> Related #https://github.com/advthreat/iroh/issues/2249
This PR implements bulk update and fixes 404 on update.

- bulk update implements the documentation [already submitted](https://github.com/threatgrid/ctia/blob/master/doc/bulk-bundle.org#delete--patch--update).
- Update was implemented as an upsert in the flow, while the update events expected only update with a [precondition of non nil previous document](https://github.com/threatgrid/ctia/blob/5be5876f63780612ea0e24a41efbf0485b9fc640/src/ctia/entity/event/obj_to_event.clj#L22).
Then trying to create a document with PUT triggers a 500, at least since the introduction of the mentionned `pre`, 2 years ago.
I chose the behavior of IROH with PUT that only update.

<a name="qa">[§](#qa)</a> QA
============================
## test bulk update
1) bulk create few entities `POST /ctia/bulk`
``` javascript
{
  "incidents": [
    {
      "description": "## Summary:\n\nOn Monday, June 15th at 3:34am GMT, a host (UUID #dc0415fe-af42-11ea-b3de-0242ac130004) on VLAN 414 established contact with a known Emotet Epoch 2 Command and Control server, triggering an event alarm..",
      "assignees": [
        "saintx"
      ],
      "type": "incident",
      "source": "Modeling Incidents in CTIM Tutorial",
      "external_ids": [
        "pr1187-c56de1c94c1ce862c4e6d9883393aacc58275c0c4dc4d8b48cc4db692bf11e4f"
      ],
      "short_description": "Incident Report: 2020-06-15 3:34am (Emotet Botnet Attack)",
      "title": "2020-06-15-0334-emotet-botnet-report",
      "incident_time": {
        "opened": "2020-06-15T03:43:27.368Z",
        "reported": "2020-06-15T03:34:36.298Z"
      },
      "discovery_method": "NIDS",
      "source_uri": "https://github.com/threatgrid/ctim/blob/master/src/doc/tutorials/modeling-incidents-in-ctim.md",
      "categories": [
        "Malicious Code"
      ],
      "status": "Containment Achieved",
      "id": "transient:ctim-tutorial-incident-c56de1c94c1ce862c4e6d9883393aacc58275c0c4dc4d8b48cc4db692bf11e4f",
      "confidence": "High"
    }
  ],
  "sightings": [
    {
      "observables": [
        {
          "type": "ip",
          "value": "98.15.140.226"
        }
      ],
      "type": "sighting",
      "source": "Modeling Incidents in CTIM Tutorial",
      "external_ids": [
        "pr1187-7b36e0fa2169a3ca330c7790f63c97fd3c9f482f88ee1b350511d8a51fcecc8d"
      ],
      "source_uri": "https://github.com/threatgrid/ctim/blob/master/src/doc/tutorials/modeling-incidents-in-ctim.md",
      "id": "transient:ctim-tutorial-sighting-7b36e0fa2169a3ca330c7790f63c97fd3c9f482f88ee1b350511d8a51fcecc8d",
      "count": 1,
      "severity": "High",
      "tlp": "green",
      "timestamp": "2020-06-15T03:34:36.298Z",
      "confidence": "High",
      "observed_time": {
        "start_time": "2020-06-15T03:34:36.298Z"
      }
    }
  ],
  "relationships": [
    {
      "type": "relationship",
      "source": "Modeling Incidents in CTIM Tutorial",
      "source_uri": "https://github.com/threatgrid/ctim/blob/master/src/doc/tutorials/modeling-incidents-in-ctim.md",
      "source_ref": "transient:ctim-tutorial-sighting-7b36e0fa2169a3ca330c7790f63c97fd3c9f482f88ee1b350511d8a51fcecc8d",
      "target_ref": "transient:ctim-tutorial-incident-c56de1c94c1ce862c4e6d9883393aacc58275c0c4dc4d8b48cc4db692bf11e4f",
      "relationship_type": "member-of",
      "external_ids": [
        "pr1187-2c1f3fcaf89d294bf7d038f470f6cb4a81dc1fad6ff5deeed18a41bf6fe14f00"
      ]
    },
    {
      "type": "relationship",
      "source": "Modeling Incidents in CTIM Tutorial",
      "source_uri": "https://github.com/threatgrid/ctim/blob/master/src/doc/tutorials/modeling-incidents-in-ctim.md",
      "source_ref": "transient:ctim-tutorial-sighting-7b36e0fa2169a3ca330c7790f63c97fd3c9f482f88ee1b350511d8a51fcecc8d",
      "target_ref": "https://ex.tld/ctia/indicator/indicator-b790ade3-e45e-48d4-7d06-f0079e6453a0",
      "description": "Sighting of host communication with known Emotet Epoch 2 C&C server",
      "relationship_type": "sighting-of",
      "external_ids": [
        "pr1187-f879541251b139dfbfbed0f5c66a7c6d20246074241fa2f814f0f3eb2250def8"
      ]
    }
  ]
}
```

2) Note all the ids and bulk update `POST /ctia/bulk` by changing the source
``` javascript
{
  "incidents": [
    {
      "id": "${incident-id}",
      "description": "bulk updated",
      "source": "bulk update test",
      "short_description": "bulk updated",
      "title": "2020-06-15-0334-emotet-botnet-report",
      "incident_time": {
        "opened": "2020-06-15T03:43:27.368Z",
        "reported": "2020-06-15T03:34:36.298Z"
      },
      "discovery_method": "NIDS",
      "source_uri": "https://github.com/threatgrid/ctim/blob/master/src/doc/tutorials/modeling-incidents-in-ctim.md",
      "external_ids": [
        "pr1187-7b36e0fa2169a3ca330c7790f63c97fd3c9f482f88ee1b350511d8a51fcecc8d"
      ],
      "categories": [
        "Malicious Code"
      ],
      "status": "Containment Achieved",
      "confidence": "High"
    }
  ],
  "sightings": [
    {
      "id": "${sighting-id}",
      "observables": [
        {
          "type": "ip",
          "value": "98.15.140.226"
        }
      ],
      "type": "sighting",
      "source": "bulk update test,
      "external_ids": [
        "pr1187-7b36e0fa2169a3ca330c7790f63c97fd3c9f482f88ee1b350511d8a51fcecc8d"
      ],
      "source_uri": "https://github.com/threatgrid/ctim/blob/master/src/doc/tutorials/modeling-incidents-in-ctim.md",
      "count": 1,
      "severity": "High",
      "tlp": "green",
      "timestamp": "2020-06-15T03:34:36.298Z",
      "confidence": "High",
      "observed_time": {
        "start_time": "2020-06-15T03:34:36.298Z"
      }
    }
  ],
  "relationships": [
    {
     "id": "${relationship-id-1}",
      "type": "relationship",
      "source": "bulk update test",
      "source_uri": "https://github.com/threatgrid/ctim/blob/master/src/doc/tutorials/modeling-incidents-in-ctim.md",
      "source_ref": "transient:ctim-tutorial-sighting-7b36e0fa2169a3ca330c7790f63c97fd3c9f482f88ee1b350511d8a51fcecc8d",
      "target_ref": "transient:ctim-tutorial-incident-c56de1c94c1ce862c4e6d9883393aacc58275c0c4dc4d8b48cc4db692bf11e4f",
      "relationship_type": "member-of",
      "external_ids": [
        "pr1187p-2c1f3fcaf89d294bf7d038f470f6cb4a81dc1fad6ff5deeed18a41bf6fe14f00"
      ]
    },
    {
      "id": "${relationship-id-2}",
      "type": "relationship",
      "source": "bulk update test",
      "source_uri": "https://github.com/threatgrid/ctim/blob/master/src/doc/tutorials/modeling-incidents-in-ctim.md",
      "source_ref": "transient:ctim-tutorial-sighting-7b36e0fa2169a3ca330c7790f63c97fd3c9f482f88ee1b350511d8a51fcecc8d",
      "target_ref": "https://ex.tld/ctia/indicator/indicator-b790ade3-e45e-48d4-7d06-f0079e6453a0",
      "description": "Sighting of host communication with known Emotet Epoch 2 C&C server",
      "relationship_type": "sighting-of",
      "external_ids": [
        "pr1187-f879541251b139dfbfbed0f5c66a7c6d20246074241fa2f814f0f3eb2250def8"
      ]
    }
  ]
}
```
3) bulk export the incident `GET /ctia/bulk?incidents=${incident-id}`and check that all entities were fully updated with `"source": "bulk update test"` for all entities and `"description": "bulk updated"` for the incident.

## test the fix of 404 on update
 
1. Try to create an incident with PUT with a non existing id, it should return a 404.
``` javascript
PUT /ctia/incident/missing-0e14cef7-fbd9-4c06-a6d6-332ad82d5b32
{
      "description": "## Summary:\n\nOn Monday, June 15th at 3:34am GMT, a host (UUID #dc0415fe-af42-11ea-b3de-0242ac130004) on VLAN 414 established contact with a known Emotet Epoch 2 Command and Control server, triggering an event alarm..",
      "source": "Modeling Incidents in CTIM Tutorial",
      "short_description": "Incident Report: 2020-06-15 3:34am (Emotet Botnet Attack)",
      "title": "2020-06-15-0334-emotet-botnet-report",
      "incident_time": {
        "opened": "2020-06-15T03:43:27.368Z",
        "reported": "2020-06-15T03:34:36.298Z"
      },
      "discovery_method": "NIDS",
      "status": "Containment Achieved",
      "confidence": "High"
    }
```
2. Create this entity, with the POST route and note the returned entity id 
``` javascript
POST /ctia/incident
{
      "description": "## Summary:\n\nOn Monday, June 15th at 3:34am GMT, a host (UUID #dc0415fe-af42-11ea-b3de-0242ac130004) on VLAN 414 established contact with a known Emotet Epoch 2 Command and Control server, triggering an event alarm..",
      "source": "Modeling Incidents in CTIM Tutorial",
      "short_description": "Incident Report: 2020-06-15 3:34am (Emotet Botnet Attack)",
      "title": "2020-06-15-0334-emotet-botnet-report",
      "incident_time": {
        "opened": "2020-06-15T03:43:27.368Z",
        "reported": "2020-06-15T03:34:36.298Z"
      },
      "discovery_method": "NIDS",
      "status": "Containment Achieved",
      "confidence": "High"
    }
```
3. update this entity (using the created id), it should return a 200.
``` javascript
PUT /ctia/incident/{incident-id}
{
      "description": "updated",
      "source": "Modeling Incidents in CTIM Tutorial",
      "short_description": "updated",
      "title": "updated",
      "status": "Open",
      "confidence": "High"
    }
```
4. Retrieve this entity and check that it was properly updated 
``` javascript
GET /ctia/incident/{incident-id}
```

<a name="release-notes">[§](#release-notes)</a> Release Notes
=============================================================

```
intern: one-line description for internal eyes only.
public: one line description for public release notes.
```

<a name="squashed-commits">[§](#squashed-commits)</a> Squashed Commits
======================================================================

### CODE REVIEW

File: src/ctia/flows/crud.clj [hunk 1]
```diff
@@ -178,16 +177,12 @@
                                             tempids
                                             identity-map)
                         :update
-                        (if-let [prev-entity (get-prev-entity entity-id)]
```
Comment (ereteog) [inline src/ctia/flows/crud.clj:181]: this is the patch for the 404.

File: test/ctia/test_helpers/crud.clj [hunk 2]
```diff
@@ -296,24 +296,14 @@
               (is (= with-updates
                      updated-record))
               (is (= updated-record
-                     stored-record)))
-
-            (when revoke-tests?
```
Comment (ereteog) [inline test/ctia/test_helpers/crud.clj:301]: moved this test from this let since it was not related.

File: test/ctia/test_helpers/crud.clj [hunk 3]
```diff
@@ -296,24 +296,14 @@
...
-                                           (format "ctia/%s/%s/expire" entity-str (:short-id record-id))
-                                           :headers headers
-                                           (when revoke-tests-extra-query-params
-                                             [:query-params revoke-tests-extra-query-params]))]
-                       (is (= 200 (:status response))
-                           (format "POST %s/:id/expire succeeds" entity-str))
-                       (is (= fixed-now (-> response :parsed-body :valid_time :end_time))
-                           ":valid_time properly reset")))))))
+                     stored-record))
+              (is (= 404
```
Comment (ereteog) [inline test/ctia/test_helpers/crud.clj:300]: the test for the 404 on update.

File: src/ctia/bulk/core.clj [hunk 4]
```diff
@@ -253,6 +278,19 @@
...
+  (apply + (map count (vals bulk))))
+
+(defn get-bulk-max-size [get-in-config]
+  (get-in-config [:ctia :http :bulk :max-size]))
+
+(defn bad-request?
+  [bulk
+   {{:keys [get-in-config]} :ConfigService}]
+  (when (> (bulk-size bulk) (get-bulk-max-size get-in-config))
+    (bad-request (str "Bulk max nb of entities: "
```
Comment (DeLaGuardo) [inline src/ctia/bulk/core.clj:291]: Looks like this line will be included in the outgoing response. Introducing non-common shorthand might be confusing. Also how about including actual bulk size into the response body together with maximum size from the config?

```suggestion
    (bad-request (str "Bulk max number of entities: "
```

File: src/ctia/bulk/routes.clj [hunk 5]
```diff
@@ -33,23 +33,34 @@
...
+             (common/created (core/create-bulk bulk
+                                               {}
+                                               login
+                                               (common/wait_for->refresh wait_for)
+                                               services))))
+       (PUT "/" []
+         :return (s/maybe (bulk.schemas/BulkActionsRefs services))
+         :summary "UPDATE many entities at once"
+         :query-params [{wait_for :- (describe s/Bool "wait for updated entities to be available for search") nil}]
+         :body [bulk (bulk.schemas/BulkUpdate services) {:description "a new Bulk Patch object"}]
```
Comment (DeLaGuardo) [inline src/ctia/bulk/routes.clj:55]:
```suggestion
         :body [bulk (bulk.schemas/BulkUpdate services) {:description "a new Bulk Update object"}]
```

**Review by DeLaGuardo — COMMENTED** [review]
I got the impression that update and patch methods effectively do the same thing. Probably I'm missing something obvious and they do serve for different reasons but maybe it is possible to combine those handlers into one? right now I can see a lot of identical code shared between them

File: src/ctia/bulk/core.clj [hunk 6]
```diff
@@ -214,7 +239,7 @@
   "
   [func bulk & args]
   (try
-    (->> bulk
+    (->> (prepare-bulk bulk)
```
[the remark this answers is not in the document]
Comment (ereteog) [reply src/ctia/bulk/core.clj:242]: it totally makes sense

**Comment by ereteog** [comment]
> I got the impression that update and patch methods effectively do the same thing. Probably I'm missing something obvious and they do serve for different reasons but maybe it is possible to combine those handlers into one? right now I can see a lot of identical code shared between them

You are totally right. 
This is the result of my iterative refactoring of the flow code to introduce bulk-delete, bulk-patch and bulk-update. The bulk flow was mainly tested at a higher level from the route tests. Then I first introduced more granular tests to transform these flow for accepting list of entities. Then I made iterative changes and "simplification" (not sure yet it's simpler ;-)), until this last iteration.
Then one more iteration is to transform these functions to simply accept the Flow structure as an argument and factorize the logic for managing previous entities. 
Once done, we could have something close to:
```
(defn patch-flow
    [flow] 
    (-> flow
          prepare-patch-flow
          patch-entities
          update-flow))
```

File: src/ctia/bulk/core.clj (hunk 6 above)
[the remark this answers is not in the document]
Comment (ereteog) [reply src/ctia/bulk/core.clj]: I also fixed few other `(->> ... (into {}))`

**Review by msprunck — APPROVED** [review]
LGTM!

File: src/ctia/bulk/core.clj [hunk 7]
```diff
@@ -244,15 +267,28 @@
...
+(defn bulk-size [bulk]
+  (reduce + (map count (vals bulk))))
+
+(defn get-bulk-max-size [get-in-config]
+  (get-in-config [:ctia :http :bulk :max-size]))
+
+(defn validate-bulk-size!
+  [bulk
+   {{:keys [get-in-config]} :ConfigService}]
+  (when (> (bulk-size bulk) (get-bulk-max-size get-in-config))
```
Comment (agzam) [inline src/ctia/bulk/core.clj:288]:
```suggestion
  (when (< (get-bulk-max-size get-in-config) (bulk-size bulk))
```

Per long lasting tradition, Lispers prefer using `<` and `<=`. "Elements of Clojure" nicely illustrates that in "Idioms" chapter:

> But the most confusing operators, by far, are the inequalities. As children, we were often taught the meaning of 3 > 1 by giving the > tiny teeth, transforming it into a hungry alligator that always wants to eat the largest number. Here the infix notation isn’t an incidental detail; it’s central to our mental model.
> Small wonder, then, that (> a b) is difficult to read. The > doesn’t point at anything; we have to move it over by one term before realizing that the hungry alligator wants to eat a. Here, too, there is a useful mnemonic: we can think of > as a downward slope, and < as an upward slope. This way, we can read (> a b c) as a descending collection of values fromatocand(< a b c)as ascending.
> Even with this mnemonic, readers can get confused. Subtraction and division, at least, have a fixed relationship between the left and right terms. With each inequality, we have to decide anew whether to use (< a b) or (> b a). Almost always, this decision
> is arbitrary; we can reduce the burden on the reader by choosing one and sticking with it.
> Inequalities should be ordered least to greatest, except where reordering terms hurts code clarity.

**Review by agzam — APPROVED** [review]

**Pull request merged**